Add settings checks

gingermusketeer opened the pull request:
Let me know what you think of this fix for #32.

Note: This removes the eval of the dot file in favor of manually parsing the config file.

Remaining changes:
- [x] Tests

gingermusketeer commented:
When config value is invalid:
<img width="504" alt="screen shot 2017-10-20 at 10 49 29 pm" src="https://user-images.githubusercontent.com/1177034/31848174-0363f516-b5e9-11e7-8b8d-8b73a339fdcd.png">
When config is unknown:
<img width="236" alt="screen shot 2017-10-20 at 10 49 15 pm" src="https://user-images.githubusercontent.com/1177034/31848175-03750298-b5e9-11e7-911e-200f0a221fd8.png">

mjago commented:
@gingermusketeer :+1: Looks good in principle. It would be good to see a benchmark as a speed increase would be great 😃  . I guess config versioning could be implemented also ref: #32.

gingermusketeer commented:
@mjago How do you normally go about benchmarking rufo? Would it make sense to have some tools for this in the codebase?

splattael commented on lib/rufo/formatter/settings.rb:
```diff
@@ -1,93 +1,29 @@
+      default = valid_options.first
+      value = options.fetch(name, default)
+      unless valid_options.include?(value)
+        STDERR.puts "invalid value for #{name}: #{value}. Valid values are: " \
```
Inspect `value` like in

```ruby
STDERR.puts "invalid value for #{name}: #{value.inspect}
```

splattael commented on lib/rufo/formatter/settings.rb:
```diff
@@ -1,93 +1,29 @@
+      value = options.fetch(name, default)
+      unless valid_options.include?(value)
+        STDERR.puts "invalid value for #{name}: #{value}. Valid values are: " \
+                    "#{valid_options.join(', ')}"
```
Inspect valid options:

```ruby
"#{valid_options.map(&:inspect).join(', ')}"
```

Before: `Valid values are: dynamic, one`
After: `Valid values are: :dynamic, :one`

splattael commented on lib/rufo/formatter/settings.rb:
```diff
@@ -1,93 +1,29 @@
-  def double_newline_inside_type(value)
-    @double_newline_inside_type = no_dynamic("double_newline_inside_type", value)
-  end
+  attr_accessor *OPTIONS.keys
```
Do we need real setters and getters?

Currently, these are just instance variables.

splattael commented on lib/rufo/formatter/settings.rb:
```diff
@@ -1,93 +1,29 @@
+                    "#{valid_options.join(', ')}"
+        value = default
+      end
+      self.send("#{name}=", value)
```
This should change to `instance_variable_set` if we want to avoid `attr_accessor` above.

splattael commented on lib/rufo/command.rb:
```diff
@@ -116,18 +116,35 @@ def format(code, dir)
   end
 
+  def parse_dot_file(file_contents)
+    file_contents.lines
```
How about using `each_with_object` and split config line in exact 2 parts?

```ruby
    file_contents.lines
                 .map { |s| s.strip.split(/\s+/, 2) }
                 .each_with_object({}) do |(name, value), acc|
      if value.start_with?(':')
        value = value[1..-1].to_sym
      elsif value == 'true'
        value = true
      elsif value == 'false'
        value = false
      else
        STDERR.puts "Unknown config value=#{value.inspect} for #{name.inspect}"
        next
      end
      acc[name.to_sym] = value
    end
```

mjago commented:
@gingermusketeer re. benchmarking I usually use a project with a large number of ruby files as the test to magnify the parsing speed value. Then in git I reset any changes made to the project by Rufo to reset the start position. One such large project is ruby/spec (5000 or so files).

gingermusketeer commented:
I did a little benchmarking on this looks to be a slight improvement of 2-4 seconds for a set of 10 runs against the ruby/spec code base.

bessey commented on lib/rufo/formatter/settings.rb:
```diff
@@ -1,93 +1,27 @@
+                     "values are: #{valid_options.map(&:inspect).join(', ')}"
+        value = default
+      end
+      self.instance_variable_set("@#{name}", value)
```
Relying on instance variables seems error prone to me, leaves us open to typoing the setting entirely in the formatter and not realising it. What about adding
```ruby
attr_accessor *OPTIONS.keys
```
... to the class, and replacing this with
```
public_send(name, value)
```
and then referring to these settings by their methods rather than instance variables in the formatter?

bessey replied:
Side note, this isn't you, but having `rufo/formatter/settings.rb` reopen an existing class and add methods to it had me very confused. How about this is instead a `Rufo::Settings` module that is mixed into `Rufo::Formatter` explicitly?

gingermusketeer commented:
@bessey Thanks for the feedback. I think that makes sense so I have gone a head and changed it. 

Your first bit of feedback is counter to something that was touched on in some earlier [feedback](https://github.com/ruby-formatter/rufo/pull/41#discussion_r146192563). I think you do have a good point about relying on instance variables and I have changed the formatter code so that we get the benefit. @splattael Be good to get your input on the last few commits.

bessey commented:
Oops, sorry to have introduced conflict. I tend to avoid @instance_vars since a typo can creep in silently, but admittedly in a code base as well tested as rufo, that may be less of a concern than it usually is to me.

Thank you for making the changes nonetheless!

gingermusketeer commented:
Going to merge this. Let me know if anything else needs addressing and I will do that in a subsequent PR.
